keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,338 @@
+    fi
+done
+```
+
+For example, on a 2-zone cluster, suppose an application is specified with
+NodeAffinity "app in [zone1, zone2]".
+
+- If "MaxSkew" is 1, rollout of its replicas can be 1/0 => 1/1 => 2/1. It can't
+  be 2/0 or 3/1.
+- If "MaxSkew" is 2, rollout of its replicas can be 1/0 => 1/1 or 2/0.
+
+This algorithm works for both Design A<sup>1</sup> and B.
+
+<sup>1</sup> As NodeAffinityTerm doens't have info like
+podSelector/namespaces/topologyKey, to make Design A work, those info need to be
+inferred based on NodeAffinity spec.
+
+#### PodAffinity
+
+```bash
+for each candidate node; do
+    if "EvenSpreading" is enabled; then
+        count number of matching pods on the topology domain this node belongs to
+        if matching num - minMatching num < defined MaxSkew; then
+            add it to candidate list
+        fi
+    elif "EvenSpreading" is disabled; then
+        keep current logic as is
+    fi
+done
+```
+
+For example, in a 3-zones cluster, the matching pods spread as 2/1/0, then
+
+- if "MaxSkew" is 1, incoming pod can only be deployed onto zone3 - i.e. 2/1/1
+- if "MaxSkew" is 2, incoming pod can be deployed onto zone2 or zone3 - i.e.
+  2/2/0 or 2/1/1
+
+This algorithm works for both Design A and B.
+
+#### PodAntiAffinity
+
+Design A provides a way to trigger "toleration" mode to broaden semantics of
+PodAntiAffinity so as to provide a solution for Story 3. The algorithm is
+roughly described as below:
+
+```bash
+# need to have a structure to know if there is at least one qualified candidate
+existCandidate = true if there is one qualified candidate globally else false
+if not existCandidate and "EvenSpreading" is disabled; then
+    return
+fi
+for each candidate node; do
+    if existCandidate; then
+        add it to candidate list if this node is qualified
+    elif "EvenSpreading" is enabled; then
+        count number of miss-matching pods on the topology domain this node belongs to
+        if misMatching# - minMisMatching# < defined MaxSkew; then
+            add it to candidate list
+        fi
+    else
+        keep current logic as is
+    fi
+done
+```
+
+> Given that if "even spreading" is applied on PodAntiAffinity, it tolerates
+> additional pods to co-locate in the same topology, hence the symmetry of
+> PodAntiAffinity is not guaranteed as well.
+
+While for Design B, it's not applicable to tweak semantics of PodAntiAffinity
+because it works independently as a predicate/priority. Hence Design B doesn't
+work for Story 3.
+
+### Pros/Cons
+
+**Pros of Design A:**
+- Less API changes
+- Less code changes (code can be built on existing InterPodPredicate, as well as
+  the internal data structures)
+
+**Cons of Design A:**
+- The support on NodeAffinity is vague
+- Current API only supports predicate
+
+**Pros of Design B:**
+- Independent design, so can work independently with Affinity API
+- Support both predicate and priority
+
+**Cons of Design B:**
+- Not working for Story 3
```

Comment:
Why doesn't it work for story 3? The user could set MaxSkew to 2 (or larger) and then it should work.

Reply:
MaxSkew won't work once PodAntiAffinity is specified. Unless we mixin the semantics of PodAntiAffinity into Design B, which IMO is not what we want.

Reply:
I would expect almost all users to use one of the two features, but I strongly agree that these should be documented well. We will deprecate and change PodAntiAffinity to be limited to node only after this feature is out.